### DISCUSSION TITLE
Always run node

### DISCUSSION THREAD

**Discussion by adieyal** [comfyanonymous/ComfyUI#40] [General]
Is it possible to mark a node to always be run, even if it hasn't changed? For example, if I have a node that generates random numbers, I would like to get a new random number every time I generate an image. Right now the graph seems to be optimised to prevent re-running of nodes that  haven't changed. 

I've looked at the KSampler node, but I don't see anything obvious that indicates that a new random seed should always be generated.

**Comment by comfyanonymous** [comment]
https://github.com/comfyanonymous/ComfyUI/blob/master/nodes.py#L763

If it's a custom node you are writing you can add an IS_CHANGED function to it. If the output of this function changes from one execution to the next the node will be run again. This is used in the LoadImage node to only rerun the graph if the image changes on the disk for example.

**Reply by adieyal** [reply]
Is it possible to keep state inside the custom node? Say I have a list of prompts that I want to cycle through. E.g.

```python
import random

class PromptNode:
    @classmethod
    def INPUT_TYPES(cls):
        return {"required": {"text": ("STRING", {"multiline": True})}}

    RETURN_TYPES = ("PROMPT",)
    FUNCTION = "generate"

    CATEGORY = "prompts"

    @classmethod
    def IS_CHANGED(cls, text):
        return random.random()

    def __init__(self):
        self._my_prompts = ["prompt1", "prompt2"]

    def generate(self, text):

        new_prompt = self._my_prompts[0]
        self._my_prompts = self._my_prompts[1:] + [new_prompt]

        return (new_prompt, )

```

I want this node to be run for every generation. 

If I set `IS_CHANGED` to something that is always different, then my `PromptNode` is re-created, losing my `self._my_prompts` list. If I don't use `IS_CHANGED` then my `generate(...)` method is only called once. I can't store the state in the class, since I might have a number of PromptNodes in my graph. I could create a state lookup in the class and use that, but I was wondering if I misunderstood how this all works.

**Reply by comfyanonymous** [reply]
Yes it gets recreated every time it is executed again.

The idea behind this is that the same submitted prompt should always generate the same image no matter what was submitted before.

**Reply by adieyal** [reply]
Understood - so it's completely stateless. 

I'm trying to integrate the [dynamicprompts](https://github.com/adieyal/dynamicprompts) library. It has a combinatorial mode which let's you generate every combination of your dynamic prompt, e.g. `{red|blue|green} {circle|square}` would generate:
red circle
red square
blue circle
blue square
green circle
green square

Given that I don't have state, I don't know which of the combinations has already been generated. Any suggestions for how to go about this?

**Reply by comfyanonymous** [reply]
There is a version of dynamic prompts implemented in the frontend:

>You can use {day|night}, for wildcard/dynamic prompts. With this syntax "{wild|card|test}" will be randomly replaced by either "wild", "card" or "test" by the frontend every time you queue the prompt. To use {} characters in your actual prompt escape them like: \{ or \}.

So what you are trying to do would likely be best done there too.

**Reply by adieyal** [reply]
The built-in template language is useful, but the dynamicprompts library is a little more refined and has a ton of features - ComfyUI is an awesome framework and I'd love to move my workflow from auto1111. Dynamic Prompts is a big part of my workflow which is which I'm trying to figure out how best to use it.

**Reply by Vespinian** [reply]
In the widgets.js you could implement a function similar to seedWidget which is used to get a new seed after every generation for the Ksampler nodes but instead the function would increment a selector/iterator property by 1.  Only problem left would be to know when you've cycled trough everything.

**Comment by adieyal** [comment]
thanks, i'll take a look

**Comment by Quasimondo** [comment]
Just in case someone is still looking for a solution - this worked for me to trigger a rerun every time a workflow is run:

```
import hashlib
import time

@classmethod
def IS_CHANGED(s, image):
        #always update
        m = hashlib.sha256().update(str(time.time()).encode("utf-8"))
        return m.digest().hex()
```

**Reply by halcyon370** [reply]
update() will return None
